src/basic/alloc-util.h

```diff
@@ -79,6 +79,52 @@ static inline void freep(void *p) {
         free(*(void**) p);
 }
 
+#define MALLOCA_THRESHOLD 256
+
+#define DECLARE_MALLOCABLE_TYPE_FULL(t, n)                      \
+        static inline void malloca_release_##n (void *p) {      \
+                void *pc = *(void **)p;                         \
+                size_t a = __alignof__ (t) * 2;               \
+                if (((uintptr_t)pc & (a - 1)) == 0)             \
+                        free(pc);                               \
+        }
+
+#define DECLARE_MALLOCABLE_TYPE(t) DECLARE_MALLOCABLE_TYPE_FULL(t, t)
+
+#define _mallocable_full_(t, n) _cleanup_(malloca_release_##n) t
+#define _mallocable_(t) _mallocable_full_(t, t)
+
+DECLARE_MALLOCABLE_TYPE(char)
+DECLARE_MALLOCABLE_TYPE(uint8_t)
+DECLARE_MALLOCABLE_TYPE(int)
+DECLARE_MALLOCABLE_TYPE_FULL(long double, long_double)
+DECLARE_MALLOCABLE_TYPE_FULL(unsigned long long, unsigned_long_long)
+
+/* malloca/malloca_release are based on address alignment.
+ * Allocations that go on the stack are aligned to "align" but not to 2*"align".
+ * Allocations that go on the heap are aligned to 2*"align".
+ */
+#define malloca(sz, align)                                              \
+        ({                                                              \
+                void *_mr_;                                             \
+                size_t _mal_ = (align);                                 \
+                size_t _mda_ = _mal_ * 2;                               \
+                size_t _mlen_ = (sz);                                   \
+                if (_mlen_ < MALLOCA_THRESHOLD)                         \
+                        _mr_ = alloca_align_offset(_mlen_, _mda_, _mal_); \
+                else                                                    \
+                        _mr_ = aligned_alloc(_mda_, _mlen_);            \
+                _mr_;                                                   \
+        })
+
+#define calloca(sz, align)                              \
```

Comment:
i think this should be named `malloca0`... To me a function name derived from `calloc()` should take two size params and do overflow checking, but this macro doesn't do that...